DiD S12 fixes

sigurdfdragon opened the pull request:
Based on irc discussion https://www.wesnoth.org/irclogs/2017/10/%23wesnoth-dev.2017-10-05.log

The objective is to prevent players from playing forever or looking up in wiki/forums to complete the campaign. And also improve dialog & feel.

Comments especially wanted on the race-specific ending dialog & Change and improve objectives commit.

GregoryLundberg commented:
I like the race-specific endings.

Having read them, seeing that you might make it several times even with the gold/income bumps and other changes, while I like the objectives changing I wonder if a bit more variation for later cycles; perhaps add a bit of humorous exasperation about the seemingly inexhaustible supply of foolish heroes on the last iteration?

sigurdfdragon commented:
@GregoryLundberg Do you mean add the humorous exasperation in dialog, objectives (or maybe both)?

GregoryLundberg commented:
I was thinking of the objectives, although, now that you ask, I really don't know where it would be best. "Your scouts report there's a queue of foolish heroes forming off in the distance." "You begin to think you'd give up all your powers just to meet one wise hero!"

sigurdfdragon commented:
Updated this PR.

Attempting to enhance the execution of the story, while maintaining compatibility with IftU & AtS.

Key commits are:
Change and improve objectives
Ensure the player doesn't play the scenario forever

Changes from 1st version
Made objectives & dialog much clearer
Increased the 'don't play forever' number
I dropped the previous difficulty changes I had as they didn't really work / not sure I agree with them / don't want to play whack-a-mole with the map until it becomes a featureless cave.

GregoryLundberg commented:
ya' know I had a whole pile of changes for DiD and think I was up to S12 and thinking about skipping it and doing a PR on the rest, when RL kicked in and I had to take time off. It's still on my fork as a  subject branch .. if you're interested.

sigurdfdragon commented:
@GregoryLundberg I may take a look, though it's very unlikely I would do anything with it before 1.13.11

sigurdfdragon commented:
Updated improve objectives to have destruction of Mal Keshar always show when it is a victory condition
Changed the 4 and up additional dialog for Mal Keshar
Tweaking of the don't play forever commit.

sigurdfdragon commented:
Updated the pr.
Allows finishing campaign by hero death on the 4th time. Has dialog and objectives to match this. (I think appropriate to make the player feel the endless night before giving the out.)

Slowly increases enemy aggression to a max of 1.0 at 7 timesForever.
Goes to an endless night epilogue if the player manages to win at 10 timesForever.
Other than maybe dropping the ai caution changes or adjusting if/when the epilogue occurs, I'm done with the pr, unless there is futher comments.

sigurdfdragon commented:
Made use of [show_if] in [gold_carryover] in ensure player doesn't play forever
tweaked difficulty settings

sigurdfdragon commented:
Added a scenario repetition counter

sigurdfdragon commented:
@ln-zookeeper Things I'm not sure about:
Major - "Ensure player doesn't play forever" This is definitely still achievable to beat the scenario 10 times, even with the aggression 1.0 addition. Not sure if it should be pushed back, hidden & require a non-obvious player action to unlock (so it would go on forever if not deliberately unlocked, or dropped completely.

Minor - Might change ai caution drop dynamics to 0.30 - 0.05 * timesForever, so it hits 0 at 6 times forever, or remove it entirely, not sure if it's doing anything significant.

ln-zookeeper commented on data/campaigns/Descent_Into_Darkness/scenarios/12_Endless_Night.cfg:
```diff
@@ -45,50 +50,89 @@
             [/gold_carryover]
 
+            [note]
+                description="<b>" + _"Forever" + " " + "$timesForever" + "</b>"
```
What's the purpose of this? You're going to have a note that says "Forever 2", "Forever 3", etc.

sigurdfdragon replied:
I put it in there so the player will easily know what iteration of the scenario they are on. Also the fact that the text of the counter is "Forever" is another clue to the player that the scenario is repeating deliberately, and I feel it goes well with the intent of the scenario.

ln-zookeeper replied:
How can a player who enters the scenario for the first time have any idea what a note that says "Forever 0" means? If they win, then on the next repetition they see "Forever 1", and won't really be any wiser. It'll most likely look more like a bug or like someone accidentally left in some kind of debug counter.

sigurdfdragon replied:
It does start at Forever 1 the first time, and would be at Forever 2 on the next repetition.

And the player probably wouldn't have any idea the first time they see it. I figured they would slowly come to the sinking realization of what it means, and that would happen by working with the rest of the changes in that commit (objectives & dialog) to creating a particular atmosphere/feel for the scenario.

I had considered using Endless Night 1, Endless Night 2... but that didn't seem flow as good to me, but maybe that is the way to go?

ln-zookeeper commented:
If you're going to have a special epilogue for surviving the maximum number of repetitions, then 10 certainly seems enough. It **should** be balanced so that it's practically impossible to get that far anyway, and thus pushing it back to 15 or something seems rather pointless.

There's a bunch of textual details I'd like to tweak, but it's easier if I just do them myself immediately after merging. Other than that, and that one odd [note] I commented on, this seems good to go.

sigurdfdragon commented:
I just bounced this with a rebase and merge on my end to see if 'Rebase and merge' would become available again. (it was greyed out) and it did become available.

sigurdfdragon commented:
I think it's difficult enough to beat it 10x with reasonable measures, the key things being when aggression hits 1.0 in Forever 7, and the income level being high enough around 9 or 10 (63/70) so that there is some constant stream of enemies that the player can't overcome.

But if the player is willing to take unreasonable/cheese/ridiculous effort, it can be cracked.

sigurdfdragon commented:
After further testing, it's still too beatable to justify the adding the alt ending, so I have removed it.
Dropped caution setting.
The difficulty adjustments for aggression at least ensure that the ai won't stand around and stare at an Ancient Lich sitting on a hex. & the income change makes it interesting and means there is a theoretical max scenario because the board would have a constant flood of enemy units. But someone could still play forever. If I make any further attempts at difficulty tweaks, they will be in a new PR.

I renamed the scenario counter to 'Endless Night'

@ln-zookeeper 
If you feel 'Endless Night doesn't work for the scenario counter, I'll remove it.
Unless there are any other specific changes, I'm done here. (for real)

CelticMinstrel commented on data/campaigns/Descent_Into_Darkness/scenarios/12_Endless_Night.cfg:
```diff
@@ -68,20 +68,59 @@
 
-            {IS_LAST_SCENARIO}
+            [note]
+                description="<b>" + _"Endless Night" + " " + "$timesForever" + "</b>"
```
Why is this split up rather than being a single translatable string? (I guess the formatting doesn't need to be translatable, but I don't see why the number shouldn't be part of the translatable string.)

sigurdfdragon replied:
1. It reuses a translated string (it's the same as the scenario title), so it doesn't add to translation work.
2. it was a habit I had gotten into with my own add-ons, to seperate formating from stuff that should be translated vs not be translated.

CelticMinstrel replied:
But what if there's a language where it makes more sense to put the number in front of the name rather than after it? I think it might be better to use the version _with_ the variable even for the title.

That said, if you'd rather just leave like this for now, I don't mind; if a translator brings up an issue with it, it can always be fixed then.

GregoryLundberg replied:
I agree with @CelticMinstrel. While it's unlikely to be a problem here, it seems to be a poor style.

On the other hand, if you include the number, you're forcing 10 (?) translations.

A more appropriate solution would be to use a string-replacement format such as `_"Endless Night %i"` ... perhaps something is available with WFL?

AI0867 replied:
You can use `_"Endless Night $timesForever"` or `_"<b>Endless Night $timesForever</b>"`. Variable substitution happens after translation.

This may still cause plural-form issues, but that's a problem we've always had in WML.

CelticMinstrel replied:
I think this probably wouldn't cause plural-form issues, but if it does they can be worked around with Lua.

ln-zookeeper replied:
If you really want a subtle counter, then the best way would to show it in the scenario name in objectives; "Endless Night (3)". However, that's not currently possible but the support would probably be easy to add.

Or, perhaps, stricken-through versions of the objective could appear per every defeated foolish hero, so on the third repetition victory objectives would look like this:

~~Defeat the foolish hero~~
~~Defeat the foolish hero~~
Defeat the foolish hero

The list might end up getting too long, though.

AI0867 replied:
> ~Defeat the foolish hero~
> ~Defeat the foolish hero~
> Defeat the foolish hero

I really like this idea.

CelticMinstrel replied:
> However, that's not currently possible but the support would probably be easy to add.

You can't include variable substitutions inside the title itself, then?

sigurdfdragon replied:
@CelticMinstrel Variable substitution doesn't work on keys directly inside [scenario]. It's mostly confined to inside events, with a few exceptions (like [story])

I think the strikethrough would only maybe work with a small, finite number of iterations of the scenario. At some point the rest of the objectives would be pushed off the screen or need a scrollbar. & I think players might find it annoying.

AI0867 commented:
> Defeat the foolish hero
> ~Defeat the foolish hero~
> ~Defeat the foolish hero~

I really like this idea.

sigurdfdragon commented:
updated to use form "<b>" + _ "Endless Night $timesForever" + "</b>" as that's what matches the convention of mainline objectives macros best.

CelticMinstrel commented:
Huh, apparently Github's markdown also parses simple HTML tags.
